java/src/main/java/ai/onnxruntime/OrtSession.java

```diff
@@ -675,6 +679,25 @@ public void registerCustomOpLibrary(String path) throws OrtException {
       customLibraryHandles.add(customHandle);
     }
 
+    /**
+     * Adds a single session configuration entry as a pair of strings.
+     *
+     * @param configKey The config key string.
+     * @param configValue The config value string.
+     * @throws OrtException If there was an error in native code.
+     */
+    public void addConfigEntry(String configKey, String configValue) throws OrtException {
+      checkClosed();
+      addConfigEntry(OnnxRuntime.ortApiHandle, nativeHandle, configKey, configValue);
+      configEntries.put(configKey, configValue);
+    }
+
+    /** Returns an unmodifiable map contains all session configuration entries. */
```

Comment:
I'd note that this a view on the config entries because future calls to `addConfigEntry` will be visible in the map returned from `getConfigEntries`, rather than a copy of it at a specific time. I think a view is the correct thing to have, but it should mention that that's what it is.

Reply:
So if a thread is iterating it and then another one is inserting, it may lead to corruption?

---
In reply to: [492834138](https://github.com/microsoft/onnxruntime/pull/5241#discussion_r492834138) [](ancestors = 492834138)

Reply:
You won't get corruption that way, just the view might not see the insert. There could be two threads inserting the same key at the same time, and depending on what the C API does in that case there could be an odd output. Making the `addConfigEntry` method synchronized is sufficient to protect that, and the overhead of doing so would be minimal as I wouldn't consider session options creation to be performance sensitive unless you're creating and throwing away a lot of models.